File: app/src/main/java/at/priv/graf/zazentimer/database/SectionDao.java
package at.priv.graf.zazentimer.database;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;
import java.util.List;

@Dao
public interface SectionDao {
    @Query("SELECT * FROM sections WHERE fk_session = :sessionId ORDER BY rank")
    List<SectionEntity> getSectionsForSession(int sessionId);

    @Query("SELECT * FROM sections WHERE _id = :id")
    SectionEntity getSectionById(int id);

    @Insert
    long insert(SectionEntity section);

    @Update
    void update(SectionEntity section);

    @Query("DELETE FROM sections WHERE _id = :id")
    void deleteById(long id);

    @Query("SELECT max(rank) FROM sections WHERE fk_session = :sessionId")
    Integer getMaxRank(int sessionId);

    @Query("UPDATE sections SET rank = :rank WHERE _id = :sectionId")
    void updateRank(int sectionId, int rank);
}

File: app/src/main/kotlin/at/priv/graf/zazentimer/database/SectionDao.kt
package at.priv.graf.zazentimer.database

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Update

@Dao
interface SectionDao {
    @Query("SELECT * FROM sections WHERE fk_session = :sessionId ORDER BY rank")
    suspend fun getSectionsForSession(sessionId: Int): List<SectionEntity>

    @Query("SELECT * FROM sections WHERE _id = :id")
    suspend fun getSectionById(id: Int): SectionEntity?

    @Insert
    suspend fun insert(section: SectionEntity): Long

    @Update
    suspend fun update(section: SectionEntity)

    @Query("DELETE FROM sections WHERE _id = :id")
    suspend fun deleteById(id: Long)

    @Query("SELECT max(rank) FROM sections WHERE fk_session = :sessionId")
    suspend fun getMaxRank(sessionId: Int): Int?

    @Query("UPDATE sections SET rank = :rank WHERE _id = :sectionId")
    suspend fun updateRank(
        sectionId: Int,
        rank: Int,
    )
}
